### PR TITLE
Add server side encryption support

### LINKED ISSUE #251: Support server-side encryption
Support server-side encryption feature of minio:
Needed headers:
- X-Amz-Server-Side-Encryption-Customer-Algorithm
- X-Amz-Server-Side-Encryption-Customer-Key
- X-Amz-Server-Side-Encryption-Customer-Key-MD5

see https://github.com/minio/cookbook/blob/master/docs/how-to-use-minio-server-side-encryption-with-aws-cli.md

### PR DESCRIPTION
**Pull request by poornas**
Fixes #251

### CODE REVIEW

File: Minio/DataModel/ServerSideEncryption.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+      {
+        headers.Add(Constants.SSEKMSKeyId, this.key);
+        headers.Add(Constants.SSEGenericHeader, "aws:kms");
+        if (context != null) 
+        {
+          headers.Add(Constants.SSEKMSContext, this.context); 
+        }
+        return;
+      }
+  }
```
Comment (harshavardhana) [inline Minio/DataModel/ServerSideEncryption.cs]: Would prefer some more comments in this file about actual implementation.

**Comment by poornas** [comment]
@harshavardhana , review comments were fixed.PTAL

File: Minio/DataModel/ServerSideEncryption.cs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+      public new EncryptionType GetType()
+      {
+        return EncryptionType.SSE_C;
+      }
+
+      public virtual void Marshal(Dictionary<string,string> headers) 
+      {
+        var md5SumStr = utils.getMD5SumStr(this.key);
+        headers.Add("X-Amz-Server-Side-Encryption-Customer-Algorithm", "AES256");
+        headers.Add("X-Amz-Server-Side-Encryption-Customer-Key",key)  ;
```
Comment (kannappanr) [inline Minio/DataModel/ServerSideEncryption.cs]: This needs to be converted into a byte array using a PBKDF function just like in `minio-go`

https://github.com/minio/minio-go/blob/master/examples/s3/put-encrypted-object.go#L64

https://github.com/minio/minio-go/blob/master/pkg/encrypt/server-side.go#L62

Comment (harshavardhana) [reply Minio/DataModel/ServerSideEncryption.cs]: Agree here, assuming that user will provide 32 byte long string is not acceptable. We should provide all the versions provided by minio-go . It's a well thought out implementation.

Comment (poornas) [reply Minio/DataModel/ServerSideEncryption.cs]: ok, changed implementation to accept a byte array as key, and validate length. c# has Aes class that generates keys, which is equivalent of PBKDF used in minio-go, this is documented in examples.

File: Minio/DataModel/ServerSideEncryption.cs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+      public virtual void Marshal(Dictionary<string,string> headers)
+      {
+        headers.Add(Constants.SSEGenericHeader, "AES256"); 
+        return;
+      }
+  }
+
+  // Server side encryption with AWS KMS managed keys
+  public class SSEKMS : SSES3
```
Comment (kannappanr) [inline Minio/DataModel/ServerSideEncryption.cs]: Why is this class derived from `SSES3`? Why can't this be derived from `ServerSideEncryption`?

File: Docs/API.md [hunk 4]
```diff
@@ -840,6 +844,7 @@ __Parameters__
 | ``fileName``  | _string_  | File to upload |
 | ``contentType``  | _string_ | Content type of the file. Defaults to " |
 | ``metadata``  | _Dictionary<string,string>_ | Dictionary of meta data headers and their values.Defaults to null.|
+| ``sse``    | _ServerSideEncryption_ | server side encryption option | Optional parameter. Defaults to null |
```
Comment (kannappanr) [inline Docs/API.md]: `server side encryption` - needs to be `Server-side encryption` for consistency

File: Docs/API.md [hunk 5]
```diff
@@ -931,6 +937,8 @@ __Parameters__
 | ``destObjectName`` | _string_ | Destination object name to be created, if not provided defaults to source object name|
 | ``copyConditions`` | _CopyConditions_ | Map of conditions useful for applying restrictions on copy operation|
 | ``metadata``  | _Dictionary<string,string>_ | Dictionary of meta data headers and their values on the destination side.Defaults to null.|
+| ``sseSrc``    | _ServerSideEncryption_ | server side encryption option for source object | Optional parameter. Defaults to null |
+| ``sseSest``    | _ServerSideEncryption_ | server side encryption option for destination object| Optional parameter. Defaults to null |
```
Comment (kannappanr) [inline Docs/API.md]: `sseSest` needs to be `sseDest`

File: Minio/DataModel/ServerSideEncryption.cs (hunk 3 above)
Comment (poornas) [reply Minio/DataModel/ServerSideEncryption.cs]: fixed

File: Docs/API.md (hunk 4 above)
Comment (poornas) [reply Docs/API.md]: done.

File: Docs/API.md (hunk 5 above)
Comment (poornas) [reply Docs/API.md]: thanks - fixed it.

**Comment by poornas** [comment]
@harshavardhana, updated the example in minio.examples to api.md as well.

**Comment by poornas** [comment]
rebased to master. PTAL @kannappanr && @harshavardhana

**Comment by poornas** [comment]
@harshavardhana and @kannappanr, PTAL

File: Minio/ApiEndpoints/ObjectOperations.cs [hunk 6]
```diff
@@ -791,11 +829,18 @@ public async Task<IObservable<DeleteError>> RemoveObjectAsync(string bucketName,
         /// </summary>
         /// <param name="bucketName">Bucket to test object in</param>
         /// <param name="objectName">Name of the object to stat</param>
+        /// <param name="sse"> Server-side encryption option.Defaults to null</param>
         /// <param name="cancellationToken">Optional cancellation token to cancel the operation</param>
         /// <returns>Facts about the object</returns>
-        public async Task<ObjectStat> StatObjectAsync(string bucketName, string objectName, CancellationToken cancellationToken = default(CancellationToken))
+        public async Task<ObjectStat> StatObjectAsync(string bucketName, string objectName, ServerSideEncryption sse = null,CancellationToken cancellationToken = default(CancellationToken))
```
Comment (kannappanr) [inline Minio/ApiEndpoints/ObjectOperations.cs]: Need space here `ServerSideEncryption sse = null,CancellationToken`

File: Minio/ApiEndpoints/ObjectOperations.cs [hunk 7]
```diff
@@ -941,9 +994,19 @@ public async Task CopyObjectAsync(string bucketName, string objectName, string d
 
             if ((copySize > Constants.MaxSingleCopyObjectSize) ||
                     (srcByteRangeSize > 0 && (srcByteRangeSize != srcStats.Size)))
-                await MultipartCopyUploadAsync(bucketName, objectName, destBucketName, destObjectName, copyConditions, copySize, meta, cancellationToken).ConfigureAwait(false);
+                await MultipartCopyUploadAsync(bucketName, objectName, destBucketName, destObjectName, copyConditions, copySize, meta,sseSrc, sseDest, cancellationToken).ConfigureAwait(false);
```
Comment (kannappanr) [inline Minio/ApiEndpoints/ObjectOperations.cs]: Need space here `meta,sseSrc,`

**Pull request merged**